### ISSUE TITLE
The code loops in training phase

### ISSUE THREAD

**Issue by marcporro** [DerekQXu/GLSearch#3]
Hello @DerekQXu,
when running the test config file, we can run GLSearch on the provided datasets, using the pretrained scalable model.
However, running it in training mode does not work. The code seems to go into a loop and get stuck.
We limited the training to just one dataset, running on the CPU and with the scalable model, but the issue persists.

The issue should be immediately reproducible, by just using the training config file.
We are running the code in the Docker with the packages specified in ```requirements.txt```.

Could you help us troubleshoot the problem?

**Comment by DerekQXu** [comment]
Hi, thanks for bringing this to our attention! I'll try my best to help!

Could you provide which line the code is looping at when you stop the process? Thanks!

**Comment by marcporro** [comment]
Well, every time we stop it it's in a different line, which suggests a quite big loop.
From what we can see, it never comes out of the while loop `while len(search_stack) != 0` in the `_forward_single_tree()` function in the backtrack file.

**Comment by DerekQXu** [comment]
Hmm... maybe playing with the below 2 flags in the `config.py` file could help: https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/config.py#L156-L157

These 2 flags control when the search should prematurely stop (since worst-case, the search takes exponential time). The stop condition can be found here: https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/layers_MCSRL_backtrack_scalable.py#L860-L865

**Comment by KaiyangLi1992** [comment]
Hello @DerekQXu, I face the same issue with @Hilicot. 

I find when I train the model, the code goes into a loop and get stuck. The reason is when the code goes into the function get_action_space_data_wrapper(self, state, is_mcsp=False)（https://github.com/DerekQXu/GLSearch/blob/4e10d5f79e7f5bd53e74628f52619a0fc10f4c80/model/OurMCS/layers_MCSRL_backtrack_scalable.py#L538-L589）,  the bds_pruned could be as follows:

![image](https://user-images.githubusercontent.com/50897450/232058041-092014ab-5ad5-4f1d-8c5e-308066830998.png)

In the same time, the state.pruned_actions could be as follows
![image](https://user-images.githubusercontent.com/50897450/232058539-fa4bf9d8-eb0b-4dd8-87aa-7f78a1a1a0dd.png)

Such that the return of function _format_action_space(bds_pruned, state) is [[],[],[]] https://github.com/DerekQXu/GLSearch/blob/4e10d5f79e7f5bd53e74628f52619a0fc10f4c80/model/OurMCS/layers_MCSRL_backtrack_scalable.py#L579

 Then the code never comes out of the while loop while (len(action_space[0]) == 0). 

Could you tell me how to avoid this issue?

**Comment by win7** [comment]
Thanks for your code. I'm trying run the code, but I have the follow issue. Could you help me please?

![Screenshot from 2023-04-17 12-32-25](https://user-images.githubusercontent.com/5606910/232564912-b624b278-baba-4190-86fd-34445f042816.png)

**Comment by DerekQXu** [comment]
@KaiyangLi1992, Thank you for the detailed error message! I will get back to you this week.

@win7, This seems like a separate issue. "Weird empty loaded dict" usually arises when the dataset is not being loaded properly. I suggest looking at this issue: https://github.com/DerekQXu/GLSearch/issues/2#issue-1616403724. If the steps there do not work, please open a new issue and let me know which dataset you are using. Thanks!

**Comment by win7** [comment]
> 

Thank you for your help.

**Comment by win7** [comment]
> @KaiyangLi1992, Thank you for the detailed error message! I will get back to you this week.
> 
> @win7, This seems like a separate issue. "Weird empty loaded dict" usually arises when the dataset is not being loaded properly. I suggest looking at this issue: [#2 (comment)](https://github.com/DerekQXu/GLSearch/issues/2#issue-1616403724). If the steps there do not work, please open a new issue and let me know which dataset you are using. Thanks!

Thank you for your help.

**Comment by DerekQXu** [comment]
@KaiyangLi1992, can you provide me with the dataset where this happens? Thanks!

I have 3 hypotheses on why the while loop gets stuck: 1) there is an issue with removing exhausted bidomains, 2) there is an issue selecting the bidomain, or 3) the `increase_degree` term is too small, so it takes many iterations to exit the while loop.

we can confirm 1) is not the issue by running the following assert statement within the while loop (the assert should be satisfied):
```
action_space_all = self._format_action_space(bidomains, state)
assert len(action_space[0]) > 0
``` 

we can confirm 2) is not the issue by running the following assert statement within the while loop (the assert should be satisfied):
```
action_space_all = self._format_action_space(self._prune_topk_bidomains(bidomains, num_bds), state)
assert len(action_space[0]) > 0
``` 

If the assert statements for 1) and 2) do not pass, there is likely a deeper issue I need to look into.

If 1) and 2) are both ok, the issue is most likely 3). A quick fix to 3) is increasing the `1.4142135623` term to a larger floating point number on [this line](https://github.com/DerekQXu/GLSearch/blob/4e10d5f79e7f5bd53e74628f52619a0fc10f4c80/model/OurMCS/layers_MCSRL_backtrack_scalable.py#L553). Please let me know if this works or not. If the problem is indeed 3), I could also try writing a patch to make [this code block](https://github.com/DerekQXu/GLSearch/blob/4e10d5f79e7f5bd53e74628f52619a0fc10f4c80/model/OurMCS/layers_MCSRL_backtrack_scalable.py#L556-L579) more efficient.
```
action_space_all = self._format_action_space(self._prune_topk_bidomains(bidomains, num_bds), state)
assert len(action_space[0]) > 0
```

Thanks again for digging into this issue!

**Comment by KaiyangLi1992** [comment]
Thanks for your reply. I just use the config the same as the config_training.py.  In addtion, I met this issue in the pretrain stage with  aids700nef dataset.  Thanks again for your suggestion, I will try to debug with these suggestions.

**Comment by KaiyangLi1992** [comment]
@DerekQXu Based on your suggestion, I have reviewed the assertions and identified the reason for the issue. 

During the execution of this loop, the bidomains are as follows:

![bidomains](https://github.com/DerekQXu/GLSearch/assets/50897450/7483b393-f062-4c1f-9d49-52e61ad4865f)

However, the state.pruned_actions is as follows:

![state.pruned_actions](https://github.com/DerekQXu/GLSearch/assets/50897450/53cabe27-c9c7-439d-839d-70060c321244)

This indicates that node 0 of G_1 and node 1 of G_2 have been mapped.

Therefore, the first item:

![first item](https://github.com/DerekQXu/GLSearch/assets/50897450/2d38d9dc-f2cc-4ce0-8f8e-8cc089e97ec3)

should not be included in the bidomains list. I believe the reason for this issue is that both bidomains and natts2bds_unexhausted contain an item that should not belong to them. Could you please help me verify if my understanding is correct? Thank you very much.

**Comment by DerekQXu** [comment]
Thank you so much for this careful debugging!

Your understanding is correct. If node 0 matched node 1 in state.pruned_actions, it should no longer exist in the bidomain.

I will look at this further.

**Comment by KaiyangLi1992** [comment]
Hello @DerekQXu. I've been trying to debug the issue again, and during the process, I discovered that bidomain originates from natts2bds_unexhausted, as shown in this section of the code: 
 https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/layers_MCSRL_backtrack_scalable.py#L540-L544  .

Additionally, based on the information provided in this section of the code: 
 https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/data_structures_buffer_scalable.py#L207-L223  , it seems that natts2bds_unexhausted is derived from bd.left - self.exhausted_v.

However, I noticed that the elements in state.pruned_actions do not necessarily have to be in self.exhausted_v. According to this section of the code: 
 https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/data_structures_buffer_scalable.py#L178-L186  , the elements in state.pruned_actions are added to self.exhausted_v only if certain conditions are met.

I believe this is the reason why there are some elements present in both state.pruned_actions and bidomain. Perhaps replacing self.exhausted_v with the left part of state.pruned_actions and self.exhausted_w with the right part of state.pruned_actions in this code:  https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/data_structures_buffer_scalable.py#L214-L215   could potentially resolve this bug.

Could you please review and confirm if my understanding is correct?

Additionally, I have another question. Could you explain the purpose of setting the variable self.exhausted_v in this project?

Thank you very much!

**Comment by DerekQXu** [comment]
Thank you for your in-depth debugging! I am sorry I did not make time to fully resolve this issue. I will try my hardest to resolve this bug this week.

**tl;dr**: Your code trace is correct, but `natts2bds_unexhausted` should be `bd.left - self.exhausted_v`. `exhausted_v` is the set of nodes in `G1` belonging to the action space, where the search had previously explored all possible matchings for said nodes. The bug likely has to do with how `exhausted_v` is created.

-------------------------------

The easiest explanation is through an example:

Say that we are at the start of search and can match the following node pairs: `S = {(1, A), (1, B), (1, C), (2, A), (2, B), (2, C), (3, A), (3, B), (3, C)}` where nodes `{1,2,3}` come from `G1` and nodes `{A,B,C}` come from `G2`. All node attributes are the same.

1. Say the DQN first pick node pair `(1, A)`.
2. Next, we continue search until it backtracks to the start of search.
3. We now pick a node pair from the following set of nodes `S \ {(1,A)}`. In this case, `pruned_actions={(1, A)}`, `natts2bds_unexhausted = {left:{1,2,3}, right:{A,B,C}}`, `exhausted_v={}`, and `exhausted_w={}`. Say the DQN picks node pair `(1, B)`
4. Next, we continue search until it backtracks to the start of search.
5. We now pick a node pair from the following set of nodes `S \ {(1,A), (1,B)}`. In this case, `pruned_actions={(1, A), (1,B)}`, `natts2bds_unexhausted = {left:{1,2,3}, right:{A,B,C}}`,  `exhausted_v={}`, and `exhausted_w={}`. Say the DQN picks node pair `(2, B)`
6. Next, we continue search until it backtracks to the start of search.
7. We now pick a node pair from the following set of nodes `S \ {(1,A), (1,B), (2, B)}`. In this case, `pruned_actions={(1, A), (1,B), (2, B)}`, `natts2bds_unexhausted = {left:{1,2,3}, right:{A,B,C}}` and  `exhausted_v={}`, and `exhausted_w={}`. Say the DQN picks node pair `(1, C)`
8. Next, we continue search until it backtracks to the start of search.
9. We now pick a node pair from the following set of nodes `S \ {(1,A), (1,B), (2, B), (1, C)}`. In this case, `pruned_actions={(1, A), (1,B), (2, B), (1, C)}`, `natts2bds_unexhausted = {left:{2,3}, right:{A,B,C}}`,  `exhausted_v={1}`, and `exhausted_w={}`. Say the DQN picks node pair `(3, B)`
10. Next, we continue search until it backtracks to the start of search.
11. We now pick a node pair from the following set of nodes `S \ {(1,A), (1,B), (2, B), (1, C), (3, B)}`. In this case, `pruned_actions={(1, A), (1,B), (2, B), (1, C), (3, B)}`, `natts2bds_unexhausted = {left:{2,3}, right:{A,C}}`,  `exhausted_v={1}`, and `exhausted_w={B}`. Say the DQN picks node pair `(2, C)`
12. etc

-------------------------------

The code assumes there should always be >= 1 pairs among `outer_product(bd_u.left, bd_u.right)` from `S \ {...}`, where `bd_u` are bidomains from `natts2bds_unexhausted`. Your earlier debugging https://github.com/DerekQXu/GLSearch/issues/3#issuecomment-1557187878 exposes `pruned_actions={(0,1)}` and `bd_u = {left:{0}, right:{1}}`. Hence, `outer_product(bd_u.left, bd_u.right) ∩ {S \ {(0,1)}} = {(0,1)} ∩ {S \ {(0,1)}} = {}` and there is a bug that violates this assumption. I'm guessing the bug has to do with `0` not being appended to `exhausted_v`.

**Comment by KaiyangLi1992** [comment]
Thank you for such a detailed explanation! I understand how the code work now. I look forward to your debugging assistance.

**Comment by DerekQXu** [comment]
Hi @KaiyangLi1992 

Can you try applying the changes below? If it works, let me know and I will push the patch. Thanks!

Explanation: a node is exhausted if we previously explored all its possible matchings in the bidomain - any nodes that have already previously been exhausted.

In `data_structures_search_tree_scalable.py`:
```
     def prune_action(self, v, w):  # , remove_nodes):
         self.pruned_actions.add_lr(v, w)
         # if remove_nodes:
         for bd in unroll_bidomains(self.natts2bds):
-            if v in bd.left:
-                assert w in bd.right
-                if len(bd.right - self.pruned_actions.l2r[v]) == 0:
+            left_valid = bd.left - set(self.exhausted_v)
+            right_valid = bd.right - set(self.exhausted_w)
+            if v in left_valid:
+                assert w in right_valid
+                if len(right_valid - self.pruned_actions.l2r[v]) == 0:
                     self.exhausted_v.add(v)
                     self.recompute = True
-                if len(bd.left - self.pruned_actions.r2l[w]) == 0:
+                if len(left_valid - self.pruned_actions.r2l[w]) == 0:
                     self.exhausted_w.add(w)
                     self.recompute = True
```
In `layers_MCSRL_backtrack_scalable.py`:
```
     def _get_empty_action_space_data(self, state):
+        state.exhausted_v = set()
+        state.exhausted_w = set()
+        state.pruned_actions = DoubleDict()
         natts2bds_unexhausted = state.get_natts2bds_unexhausted(with_bids=True)
         action_space_data = \
             ActionSpaceData(self._get_empty_action_space(), natts2bds_unexhausted, None)
```

**Comment by KaiyangLi1992** [comment]
Thank you for your assistance! It has been incredibly helpful in resolving the previous bug. However, it seems that I have encountered a new bug. Based on the code's print output, the program enters into a loop after the 300th iteration. I am currently attempting to reproduce this bug.

Here is a partial printout of the program:

> Iter: 035 (35), Loss: 6.9001565		7.510 secs
 		 5.925 secs
	Iter: 040 (40), Loss: 2.7698963		5.925 secs
 		 5.581 secs
	Iter: 045 (45), Loss: 2.1779363		5.581 secs
 		 6.037 secs
	Iter: 050 (50), Loss: 7.1942840		6.037 secs
 		 5.506 secs
	Iter: 055 (55), Loss: 3.0188451		5.506 secs
 		 6.351 secs
	Iter: 060 (60), Loss: 2.7091000		6.351 secs
 		 29.000 mins, 34.189 secs
	Iter: 065 (65), Loss: 9.9412584		29.000 mins, 34.189 secs
 		 30.000 mins, 17.679 secs
	Iter: 070 (70), Loss: 5.9495721		30.000 mins, 17.679 secs
 		 27.000 mins, 25.507 secs
	Iter: 075 (75), Loss: 3.0151932		27.000 mins, 25.507 secs
 		 27.000 mins, 17.460 secs
	Iter: 080 (80), Loss: 4.5808411		27.000 mins, 17.460 secs
 		 1.000 min, 20.627 secs
        .......
        Iter: 290 (290), Loss: 5.5803185		6.235 secs
 		         4.857 secs
	        Iter: 295 (295), Loss: 1.9901568		4.857 secs
 		         5.566 secs
	        Iter: 300 (300), Loss: 3.1710000		5.566 secs


In some iterations, such as Iteration 60, the code takes 30 minutes to complete, while usually one iteration only takes around 5 seconds. I would like to know if it is normal for the pretrain phase to occasionally consume 30 minutes per iteration.

Additionally, I have discovered that the parameter "recursion_threshold" can control the iteration time. Can I modify this parameter to reduce the time it takes to reproduce the bug? If so, would it be acceptable to change the parameter from -1 to 1000?

**Comment by DerekQXu** [comment]
I am not sure about the 300th pair. We will try reproducing on our end as well. Iterations take much longer on graph pairs that are very dense, which is expected behavior.

Yes, you can safely modify the `recursion_threshold` parameter to reproduce the bug.

**Comment by KaiyangLi1992** [comment]
If you try to reproduce it on your own terminal, that would be great, and I would be very grateful. Currently, I am using the dataset "syn:np_tr=20,np_te=20,nn_core=-1,nn_tot=16,ed=5,gen_type=BA" with a V100 GPU. Is it too long for an iteration to take half an hour?

**Comment by DerekQXu** [comment]
Could you provide me the `config.py` used to reproduce the results above? I tried using `syn:np_tr=20,np_te=20,nn_core=-1,nn_tot=16,ed=5,gen_type=BA`, and every iteration ran in ~5 seconds. Thanks!

**Comment by KaiyangLi1992** [comment]
I just use the same config file as https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/config_training.py. I have not changed anything except the name of this file.

**Comment by DerekQXu** [comment]
I think the 30 minutes is expected, given the current config.

I suggest applying one of the two below fixes to speed up training:
1) set `recursion_threshold = 1` on https://github.com/DerekQXu/GLSearch/blob/main/model/OurMCS/config_training.py#L240

2) Apply the below patch in `layers_MCSRL_backtrack_scalable.py`:
```
@@ -862,7 +874,9 @@ class MCSRLBacktrack(nn.Module):
         recursion_thresh = (
                                        self.recursion_threshold is not None) and recursion_count > self.recursion_threshold
         timout_thresh = self.total_runtime is not None and timer.get_duration() > self.total_runtime
-        return recursion_thresh and timout_thresh
+        return recursion_thresh or timout_thresh
```

Hope it helps!

**Comment by KaiyangLi1992** [comment]
Hi. @DerekQXu . I have found that changing this parameter recursion_threshold  indeed helps to avoid the issue of an infinite loop, but I believe the program itself should not get stuck in such a long loop. So there might still be a bug. 

Could you please run the code again on your end? If the program runs properly on your end, it might be an issue with my environment. I will strictly follow your requirement file to reproduce the environment.

**Comment by DerekQXu** [comment]
@KaiyangLi1992, Thank you so much for confirming `recursion_threshold` fixes it! 

It did not loop on my end when I applied either of the 2 fixes from https://github.com/DerekQXu/GLSearch/issues/3#issuecomment-1601956977.

To make future setups easier, I pushed the earlier patch from https://github.com/DerekQXu/GLSearch/issues/3#issuecomment-1594046071 and the second fix from https://github.com/DerekQXu/GLSearch/issues/3#issuecomment-1601956977.  I pulled the updated code, reran with `config_training.py`, and did not see any looping behavior. Please let me know if it works on your end, and we thank you again for your debugging efforts!
